aten/src/ATen/core/Generator.h

```diff
@@ -1,24 +1,93 @@
 #pragma once
 
-#include <ATen/core/ATenGeneral.h>
 #include <stdint.h>
+#include <memory>
+#include <mutex>
+#include <deque>
+#include <atomic>
+#include <typeinfo>
+#include <utility>
+
+#include "c10/util/Exception.h"
+#include <c10/util/C++17.h>
+#include "ATen/core/Backend.h"
+#include "ATen/core/PhiloxRNGEngine.h"
+
+#if !C10_MOBILE
+#include "ATen/detail/CUDAHooksInterface.h"
+#endif
+
+/*
+* Generator note.
+* A Pseudo Random Number Generator (PRNG) is an engine that uses an algorithm to 
+* generate a seemingly random sequence of numbers, that may be later be used in creating 
+* a random distribution. Such an engine almost always maintains a state and requires a
+* seed to start off the creation of random numbers. Often times, users have
+* encountered that it could be beneficial to be able to create, retain, and destroy 
+* generator states and also be able to have control over the seed value.
+*
+* A Generator in ATen gives users the ability to read, write and modify a PRNG engine.
+* For instance, it does so by letting users seed a PRNG engine, get/set the state of the
+* engine, etc.
+*
+* By default, there is one generator state per device, and a device's generator state is 
+* lazily created. A user can use the torch.Generator() api to create their own generator.
+*/
 
 namespace at {
 
+struct Generator;
+
+/*
+* A GeneratorState object contains a generator engine and other state variables.
+* It also has copy and assign constructors so that generator states can be deep copied.
+*/
+struct CAFFE2_API GeneratorState {
+  int64_t device = -1;
+  DeviceType device_type;
+  uint64_t current_seed = 67280421310721;
+  uint64_t cuda_philox_offset_per_thread;
+  at::Philox4_32_10 cpu_engine;
+};
+
+namespace detail {
+
+// API (for internal use)
+CAFFE2_API GeneratorState createGenerator(DeviceType device_type=kCPU, int64_t device=-1);
+CAFFE2_API Generator& getDefaultGenerator(DeviceType device_type=kCPU, int64_t device=-1);
+CAFFE2_API Generator* checkGeneratorWithDefault(Generator* expr, Generator* defaultValue);
+
+} // namespace detail
+
 struct CAFFE2_API Generator {
```

Comment:
So, in your patch, what is a Generator, semantically? I'm not really sure what the old semantics were, but my vague intuition about it was that it represented both (1) a particular random number generator implementation (e.g., you could have an instance that was a CPU generator, or an instance that was a CUDA generator), and (2) the actual RNG state associated with this.

But this does not seem to be what the new class does, because you've got both getCPUEngine as well as incrementPhiloxOffset. So I'm very confused about what Generator is actually supposed to be.

Reply:
In your PR description, you say:

> Generator: An object which manages the state of the algorithm that produces pseudo random numbers. For instance, Generator class in ATen is responsible for handling seeds, creation of generators, etc.

What I am specifically asking about, is clarification of the relationship of Generator to specific device types.

Reply:
I suppose you do say:

> aten/src/ATen/core/Generator.h: Re-defines backend API for Generators. Unifies CPU and CUDA Generator structs, i.e. GeneratorState is equivalent to the union of THCGenerator and THGenerator.

The question I want to ask is, is this actually desirable. For example, we also have a HIP device type, and plans to add an FPGA device type. It seems a bit troublesome if we have to keep adding more and more fields to GeneratorState for every device type we want to add. So I'm not too happy to see this occur.

Reply:
I think we want some sort of base class, with subclasses for CPU generator and CUDA generator. One awkward thing is I'm not sure if we should have separate state and generator (as before), or maybe we should just do away with state and a Generator is just both algorithm and state. Maybe if you draft up a proposal and post it here before you code it up, it will be easier to assess.

I do agree that we don't need codegen here, let's just write out the subclasses by hand (whatever they are.)

Reply:
![_blank uml](https://user-images.githubusercontent.com/8906225/49277910-0d1ab480-f438-11e8-9bd5-30a44b0f213a.png)
How does this look? I'm not 100% sure on getState(), setState() on the subclasses, and am thinking that I may be able to make use of std::tuple, and then use them in `aten/src/TH/generic/THTensorRandom.cpp` for the getRNGState and setRNGState which deals with torch.ByteTensor.

Reply:
Hmmm.. I may not even need getState/setState for getRNGState/setRNGState tensors if the classes are copyable. Currently I am making the Generator class copyable using the constructor forwarding and grabbing the mutex before copying.

Reply:
This looks plausible. Some questions I have:

- What is the type of getState()/setState() in the top level virtual?
- Why does CPUGenerator have getCPUEngine()/setCPUEngine(), but no one else has something similar? Maybe really this is just a getEngine()/setEngine() thing?
- I'm not exactly sure about the semantics of the methods here, but I guess we can always refine in the future